### PR TITLE
Typesafe marking and general refactoring

### PR DESCRIPTION
**Pull request by MattWindsor91**
**NOTE**: This is based on `mw-uninterpret`: that should be merged first?

This mainly refactoring PRQ:
- Removes `Unmarked`: every instance where a variable could be unmarked now has this reflected in the variable type (`Var` vs `MarkedVar`);
- Makes all of the backends use the same `MarkedVar` to string converter: this is the one previously used only for HSF.  This means Z3 output will use `VfooA` instead of `foo!after`;
- Accidentally put the heap-explicit version of the lclist in with the above… oops.

### CODE REVIEW

File: ExprTests.fs [hunk 1]
```diff
@@ -1,166 +0,0 @@
-/// <summary>
```
Comment (mjp41) [inline ExprTests.fs]: Where did all these tests go?

**Comment by mjp41** [comment]
Okay. Looks good to me.  Just two small things

I am having a slight issue running NUnit I get exceptions for
- Starling.Core.GuardedView+Tests+NUnit
- Starling.Tests.Instantiate+InstantiateTests
- Starling.Tests.Optimiser+OptimiserTests

One full exception follows:

```
SetUp failed for test fixture Starling.Tests.Optimiser+OptimiserTests
An exception was thrown while loading the test.
System.NullReferenceException: Object reference not set to an instance of an object.
   at NUnit.Framework.Internal.AsyncInvocationRegion.IsAsyncOperation(MethodInfo method)
   at NUnit.Framework.Internal.Builders.NUnitTestCaseBuilder.CheckTestMethodSignature(TestMethod testMethod, TestCaseParameters parms)
   at NUnit.Framework.Internal.Builders.NUnitTestCaseBuilder.BuildTestMethod(IMethodInfo method, Test parentSuite, TestCaseParameters parms)
   at NUnit.Framework.TestCaseSourceAttribute.<BuildFrom>d__16.MoveNext()
   at NUnit.Framework.Internal.Builders.DefaultTestCaseBuilder.BuildFrom(IMethodInfo method, Test parentSuite)
   at NUnit.Framework.Internal.Builders.NUnitTestFixtureBuilder.AddTestCasesToFixture(TestFixture fixture)
   at NUnit.Framework.Internal.Builders.NUnitTestFixtureBuilder.BuildFrom(ITypeInfo typeInfo)
   at NUnit.Framework.Internal.Builders.DefaultSuiteBuilder.BuildFrom(ITypeInfo typeInfo)
NUnit Adapter 3.0.9.0 executing tests is finished
```

Minor nit, it would be good if stuff didn't disappear in one commit, and then appear in a later one.  Makes it harder review.

**Comment by MattWindsor91** [comment]
@mjp41 I have an awful feeling I've accidentally pushed a version with an infinite loop in it.  If it's the same one I had earlier this week I've already fixed it.  I keep forgetting I'm not in a lazy pure language anymore…

Apologies, this entire pull request is a total disaster.  Was hoping I wouldn't have messy PRQs again, but the problem is I keep trying to make minor changes, then realising they snowball into bigger changes, then keep changing my mind several times about the implementation mid-flight, then before I know it I've spent 3 days on one commit and can't easily break it up.

**Comment by mjp41** [comment]
Can you add that fix to this pull request?  So I can put it into master.

**Comment by mjp41** [comment]
@CaptainHayashi  Can you add the fix? Will this affect the test cases that are failing for me?

**Comment by MattWindsor91** [comment]
Hmm, on second look `mw-typesafe-mark` is passing 100% for me…  Odd

EDIT: This is after checking out `origin/mw-typesafe-mark` and xbuilding, so it _should_ be the same branch as on here.

**Comment by septract** [comment]
NUnit is passing 100% for me too. What's the best way to narrow this down?

**Comment by mjp41** [comment]
So can you confirm it is actually running the test "GFunc with no guard and no parameters has no variables".  On mine it is failing to even find this test with a NPE, and then happily reports success.

**Comment by mjp41** [comment]
So there is also an issue with this PRQ and the hsf output.  It generates the following at the top for -s hsf for the Ticket Lock example (before the query_naming secion):

```
emp(serving, ticket) :-
    serving = 0,
    ticket = 0.
v_holdLock_holdLock(serving, ticket) :-
    false.
false :-
    v_holdLock_holdLock(serving, ticket).
```

But should generate

```
emp(Vserving, Vticket) :-
    Vserving = 0,
    Vticket = 0.
v_holdLock_holdLock(Vserving, Vticket) :-
    false.
false :-
    v_holdLock_holdLock(Vserving, Vticket).
```

I think it is using the same map as the `query_naming` for the View Definitions.  

https://github.com/septract/starling-tool/blob/mw-typesafe-mark/Horn.fs#L305-L313

**Comment by MattWindsor91** [comment]
@mjp41 Ah, good catch… maybe we need some HSF regression tests at some point.

Am trying to find out how to kick NUnit into telling me whether it's run the offending test.  I've had this issue with it before when it gives me failures on Windows and passes on Mono.

**Comment by MattWindsor91** [comment]
@mjp41 Ok, I can't work out whether my tests are running it.  All I can say is that I'm getting 157 passes.

On the test you mentioned, I've noticed the type of some of the incoming test data is ambiguous, so I've tightened it up and will be pushing soon.  I'll look at the other failing tests to see what sticks out.  I don't know if this fixes it, but if nothing else I can check on my desktop tonight.

**Comment by mjp41** [comment]
So I have a 123 tests run.  So it is some difference between Mono and Windows. 

That didn't help either.

**Comment by MattWindsor91** [comment]
Hmm… It is definitely running the tests on my end, because commenting out the GFunc test dropped it to 156.

**Comment by mjp41** [comment]
@CaptainHayashi do you think it could related to 
https://github.com/nunit/nunit/issues/1346

**Comment by mjp41** [comment]
So changing

```
 member this.``testVarsInGFunc`` gf =
```

with

```
 member this.``testVarsInGFunc`` (gf : MGFunc) =
```

Fixes the first one.  It had inferred a generic type for the gf, the internals were not pinned down.

**Comment by mjp41** [comment]
This commit fixes the tests on Windows.  I think we can merge now.

**Pull request merged**